### PR TITLE
MRPHS-3405: Extend AWS Support

### PR DESCRIPTION
**Pull request by nirav1007**
**JIRA Id**
[MRPHS-3405](https://apporbit.atlassian.net/browse/MRPHS-3405)

**Problem**
Libretto does not support for getting Regions, AvailabilityZones, VPCs, Subnets, SecurityGroups, and for adding and removing rules to SecurityGroup & Volumes to VM.

**Solution** 
Support has been extended by implementing all these functions, which are also listed below. For various get calls (region, zone, etc.) new structures have been created and list of these structs returned for those calls.
For adding and revoking rules to SecurityGroup & adding and deleting volumes to VM, VM struct has been modified to include list of local structs created for SecurityGroup and Volume.
Also, wait functions have been implemented for Start, Halt & Destroy operations, which only returns when state of VM becomes stable.
- GetRegionList
- GetAvailabilityZoneList
- GetVPCList
- GetSubnetList
- GetSecurityGroupList
- GetImageList
- AuthorizeSecurityGroup
- RevokeSecurityGroup
- CreateVolume
- AttachVolume
- DetachVolume
- DeleteVolume  

**Testing**
Unit tested all these calls with client implemented in Halo. Logs attached.
[c3.log](https://github.com/apporbit/libretto/files/1465947/c3.log)

### CODE REVIEW

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: Timeout while waiting for VM to be in running state

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: I think we should have a separate error message here and return. Also split it into separate cases (one each for shutdown and terminated). No magic like this -- have a proper error message. "Machine expectedly terminated while waiting for it to start"...

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: i --> ii everywhere in other functions as well...

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: second is too less i guess for these operation... it should start from 10 seconds at-least.. and we should cap it at 15 mins or so in total... and maxretries should be derived from a maxtimeout (set to 15mins) constant instead of other way around. maxtimeout should be a global which can be changed easily in future if required.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: Timeout while waiting for volume to be ready.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: should have a global constant for the timeout and number of retries should be based on that timeout... Also i think 5 seconds gap between retries is okay -- we may not need exponential timeouts...

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: we can have a constant timeout instead of exponential... gap between retries 30 seconds is good i think.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: we can have a constant timeout instead of exponential... gap between retries 30 seconds is good i think.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: we can have a constant timeout instead of exponential... gap between retries 30 seconds is good i think.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: same as below

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: same as below..

File: virtualmachine/aws/vm.go
Comment (ashpuri) [inline virtualmachine/aws/vm.go]: do we need to wait till ready or something else?

File: virtualmachine/aws/vm.go
Comment (ashpuri) [inline virtualmachine/aws/vm.go]: could you print volume id as well?

File: virtualmachine/aws/vm.go
Comment (ashpuri) [inline virtualmachine/aws/vm.go]: add volume id also in the error message.

File: virtualmachine/aws/vm.go
Comment (ashpuri) [inline virtualmachine/aws/vm.go]: do we need to wait for anything here?

File: virtualmachine/aws/util.go
Comment (ashpuri) [inline virtualmachine/aws/util.go]: put instance id in the message

File: virtualmachine/aws/util.go
Comment (ashpuri) [inline virtualmachine/aws/util.go:342]: do we need else here? can we have both permissions?

Comment (nirav1007) [reply virtualmachine/aws/util.go:342]: Yes, because we can not specify permission for both , ipv4Blocks and ipv6Blocks, in same block. AWS sdk will throw error in that case. Should this be mentioned in comment?

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Have not separated conditions as ShuttingDown is state leading to Terminated in AWS. So ShuttingDown is as good as Terminated. Added return statement with appropriate error message (at other places also).

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Done.

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Done.

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: For three wait functions for VM opertions viz. Running, Stopped and Terminated, have set Timeout as 15 mins and retry interval as 15 seconds as Start operation on VM brings back it to Running state in around 15 seconds and Stop and Terminate operations finish in around 63 seconds. So 15 secs interval gives optimum time for all three operations.

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Have constants for Timeout and retry Interval. Volume wait functions is made generic to match any given volume state. Timeout is set to 10 mins and retry Interval to 5 secs as all operations take around 5 secs to complete.

File: virtualmachine/aws/vm.go
Comment (nirav1007) [reply virtualmachine/aws/vm.go]: Added wait for Volume to get attached before return and change its state to In-Use.

File: virtualmachine/aws/vm.go
Comment (nirav1007) [reply virtualmachine/aws/vm.go]: Actually function name should have been waitUntilVolumeAvailable, because stable state for volume after creation and when not being attached/used by VM is Available. But now have made the wait function for volume generic to match any given state.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: Shoudn't it be just: VolInterval* time.Second?

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: could you move it just above the loop?

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: is it possible to combine all VM state related functions into single function? For simplicity of use we can create wrappers over this common function but the state checking etc. and the loop could be common.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: IMO, sleep should just be VmOpsInterval in each loop in this and all other functions. as this will go much beyond the VmOpsTimeout

File: virtualmachine/aws/vm.go
Comment (ashpuri) [inline virtualmachine/aws/vm.go]: IMO, we should create a simple wrappers over waitUntilVolumeState and call them waitfordetach/ waitforattach etc and leave the actual state name dependency in the wait.go file itself.

**Review by ashpuri — COMMENTED** [review]
Some comments inline.. get it reviewed from @avnish30jn or @phanikumarpbv as well

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Yes it should be Interval * seconds. Thanks for pointing this out.

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Yes it should be. Thanks for pointing out.

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go]: Have combined three wait functions, for Start, Stop and Terminate VMs. Have kept CreateVM wait functions as it is because it returns error with details of any missing parameter such as PublicIP, SubnetId etc which are mandatory for VM. So because of this its structure is not same as other three which we have implemented.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: comment is not true -- here and in similar function below

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go]: is this a bug/ panic situation? It seems so for me... Caller is asking for wait for state without giving any valid state

**Review by ashpuri — APPROVED** [review]
LGTM. Please wait for one more approval before merging this.

File: virtualmachine/aws/wait.go
Comment (ashpuri) [inline virtualmachine/aws/wait.go:200]: should add break

**Review by ashpuri — APPROVED** [review]

File: virtualmachine/aws/wait.go
Comment (avnish30jn) [inline virtualmachine/aws/wait.go:12]: If these constants are not cloud specific, should we move them to ```virtulamachine/.....``` ?

**Review by avnish30jn — APPROVED** [review]
LGTM, one minor comment - should we capitalize all the functions which are being used in aws package only ?

File: virtualmachine/aws/wait.go
Comment (nirav1007) [reply virtualmachine/aws/wait.go:12]: Checked in GCP and VSphere cloud, that wait functions are differently implemented in different clouds. For ex in VSphere I couldn't find any timeout used, in GCP there is wait function implemented but timeout is set to 180 seconds only.
So considering that some clouds may have wait functions in-built with the operation call, or difference in timeout and retry duration required across clouds, its better to specify it cloud specific only.

**Comment by nirav1007** [comment]
Included Filters, used for GetSubnetList, GetSecurityGroupList & GetImageList in VM struct itself because in this way we can avoid sending Filters separably as arguments to Halo & Halo-to-Liberatto. Please review the changes.

**Pull request merged**
